shared/metrics/src/main/java/io/pravega/shared/MetricsNames.java

```diff
@@ -229,4 +229,27 @@ public static String failMetricName(String metricName) {
             return metricName + "_fail";
         }
     }
+
+    /**
+     * Generate a dot separated combination of metric name and tag value[s].
+     * e.g. metric "pravega.segmentstore.container.append_count" with tag "container=6" will get:
+     * "pravega.segmentstore.container.append_count.6".
+     *
+     * @param metric the metric name.
+     * @param tags the tag(s) associated with the metric.
+     * @return the dot separated combination of metric name and tag value(s).
+     */
+    public static String nameFromTags(String metric, String... tags) {
```

Comment:
II believe that Micrometer has this same logic, why would you need to do this here?  See [Concepts -> Naming meters](http://micrometer.io/docs/concepts#_naming_meters).

Reply:
This is to keep the backwards compatibility. 
Micrometer does generate id which is a combination of metric name and tags, something like "pravega.segmentstore.container.append_count.containerid=6" (note both tag key and tag value are showing on id); however, our existing name convention is to append tag value only to metric name, something like: "pravega.segmentstore.container.append_count.6".

Reply:
Simply put, `nameFromTags` should not be used for InfluxDB; the metric name should be `pravega.segmentstore.container.append_count` in the InfluxDB scenario, because that is the effective name that the (obsolete) telegraf script would have produced.

Reply:
Additionally, in order not to keep strong reference to metric objects, Pravega always use String key to lookup a metric object; so we have to always find a way to form a unique key for a metric object.

Micrometer holds weak reference to metric object, and encourage application code to hold direct reference to metric object, which gives two benefits:
By application holding direct strong reference to metric objects, there is no need to cache metric object;
By metrics registry holding weak reference to metric objects, there is no need to manage the life-cycle of metric objects separately (like the eviction we do today), because metric objects stay with application objects, their life-cycle should be managed naturally together.

We'll probably re-factoring the metrics further in the future to make full advantage of Micrometer.

Reply:
Yes Eron you're right, when we still use flat name (which needs telegraf agent to map it back to dimensional metrics", the flat name is like this: "pravega.segmentstore.container.append_count.6.Counter"

With next PR to convert flat metric to dimensional, it will be like this:
"pravega.segmentstore.container.append_count", with a tag "containerid=6".

There are two separate keys: one is lookup key used by application to locate a metric object; another is the register key ("effective key" in your term) used by metric registry.
We always need nameFromTags() to form application lookup key; however, when there's tag available, we only use the original key (effective key) to register metric, lookup key is not used.